### PR TITLE
Adding/Upgrading Organization Resource 

### PR DESCRIPTION
**Pull request by yashrajbothra**
Signed-off-by: Yash Bothra <yashrajbothra786@gmail.com>

#### Short description of what this resolves:
This will Add some Updates related to Standard and FHIR Organization Resource

#### Changes proposed in this pull request:
1. Changed Names of Existing functions to introduce the new ones and maintain our Standard API flow.(https://github.com/openemr/openemr/commit/ae52257ac83c66fdbcde6394db0231ff66012283)

### CODE REVIEW

File: sql/database.sql [hunk 1]
```diff
@@ -1611,6 +1612,7 @@ CREATE TABLE `facility` (
   `oid` VARCHAR(255) NOT NULL DEFAULT '' COMMENT 'HIEs CCDA and FHIR an OID is required/wanted',
   `iban` varchar(50) default NULL,
   `info` TEXT,
+  UNIQUE KEY `uuid` (`uuid`),
   PRIMARY KEY  (`id`)
```
Comment (bradymiller) [inline sql/database.sql:1616]: @yashrajbothra , A common issue that happens when adding a new column to the facility table is that it breaks the install (we have all made this mistake; some of use including myself even more than once :) ). Just need to remember to place NULL for the `uuid` below this where the facility insert is (line 1623).

File: tests/Tests/Fixtures/BaseFixtureManager.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        return $idValue;
+    }
+    
+    /**
+     * Returns an unregistered/unlogged UUID for use in testing fixtures
+     * @return uuid4 string value
+     */
+    public function getUnregisteredUuid()
+    {
+        $uuid4 = Uuid::uuid4();
```
Comment (bradymiller) [inline tests/Tests/Fixtures/BaseFixtureManager.php:64]: Note that the follow will also return unregistered uuid not based on a table:
```(new UuidRegistry(['disable_tracker' => true]))->createUuid()```

**Comment by yashrajbothra** [comment]
image

Should we add an active field in table or just skip the POST & PATCH for FHIR as its never mandatory?

**Comment by bradymiller** [comment]
hi @yashrajbothra , I think the `Service Location` toggle in the ui will work for this, which maps to `service_location` in the table.

**Comment by bradymiller** [comment]
weird, it looks like my comments are anchored to the code. will try again.

File: sql/database.sql [hunk 3]
```diff
@@ -1611,14 +1612,15 @@ CREATE TABLE `facility` (
...
 ) ENGINE=InnoDB AUTO_INCREMENT=4;
 
 --
 -- Inserting data for table `facility`
 --
 
-INSERT INTO `facility` VALUES (3, 'Your Clinic Name Here', '000-000-0000', '000-000-0000', '', '', '', '', '', '', NULL, NULL, 1, 1, 0, NULL, '', '', '', '', '', '','#99FFFF','0', '', '1', '', '', '', '', '', '', '', '');
+INSERT INTO `facility` VALUES (3, 'Your Clinic Name Here', '000-000-0000', '000-000-0000', '', '', '', '', '', '', NULL, NULL, 1, 1, 0, NULL, '', '', '', '', '', '','#99FFFF','0', '', '1', '', '', '', '', '', '', '', '',NULL);
```
Comment (bradymiller) [inline sql/database.sql:1623]: note uuid is the second column, so this should be:
```
INSERT INTO `facility` VALUES (3, NULL, 'Your Clinic Name Here', '000-000-0000', '000-000-0000', '', '', '', '', '', '', NULL, NULL, 1, 1, 0, NULL, '', '', '', '', '', '','#99FFFF','0', '', '1', '', '', '', '', '', '', '', '');
```

File: tests/Tests/Fixtures/BaseFixtureManager.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        return $idValue;
+    }
+    
+    /**
+     * Returns an unregistered/unlogged UUID for use in testing fixtures
+     * @return uuid4 string value
+     */
+    public function getUnregisteredUuid()
+    {
+        return (new UuidRegistry(['disable_tracker' => true]))->createUuid();
```
Comment (bradymiller) [inline tests/Tests/Fixtures/BaseFixtureManager.php:64]: note this returns a binary uuid

**Comment by yashrajbothra** [comment]
Thanks for the Review @bradymiller :)

Wired one test is failing only in **Nginx Php 7.3** :(

File: _rest_routes.inc.php [hunk 5]
```diff
@@ -417,10 +416,20 @@
...
     "GET /fhir/Organization/:id" => function ($id) {
-        return (new FhirOrganizationRestController(null))->getOne($id);
+        return (new FhirOrganizationRestController())->getOne($id);
+    },
+    "POST /fhir/Organization" => function () {
+        RestConfig::authorization_check("admin", "users");
+        $data = (array) (json_decode(file_get_contents("php://input"), true));
+        return (new FhirOrganizationRestController())->post($data);
+    },
+    "PATCH /fhir/Organization/:id" => function ($id) {
```
Comment (bradymiller) [inline _rest_routes.inc.php:429]: just noted this. for the above 2 endpoints should have:
```
RestConfig::authorization_check("admin", "super");
```
(then will be consistent with the api permissions)

**Comment by bradymiller** [comment]
Gonna bring this in (doing a quick merge first)

**Comment by bradymiller** [comment]
Merge done. Now waiting for the travis green light :)

**Comment by bradymiller** [comment]
interesting, it looks like 1 of 5 tests or so are failing in nginx environments. guessing a sample data issue. gonna do bunch of local tests to see if can find it.

File: sql/5_0_2-to-6_0_0_upgrade.sql [hunk 6]
```diff
@@ -740,13 +740,13 @@ CREATE UNIQUE INDEX `uuid` ON `users` (`uuid`);
 ALTER TABLE `uuid_registry` ADD `table_vertical` varchar(255) NOT NULL DEFAULT '';
 #EndIf
```
Comment (bradymiller) [inline sql/5_0_2-to-6_0_0_upgrade.sql:742]: This block needs to be added back:
```
#IfMissingColumn facility_user_ids uuid
ALTER TABLE `facility_user_ids` ADD `uuid` binary(16) DEFAULT NULL;
#EndIf

#IfUuidNeedUpdateVertical facility_user_ids uid:facility_id
#EndIf

#IfNotIndex facility_user_ids uuid
CREATE INDEX `uuid` ON `facility_user_ids` (`uuid`);
#EndIf
```

Comment (yashrajbothra) [reply sql/5_0_2-to-6_0_0_upgrade.sql]: Will never do Merge's on Github Confuses me a lot :confused:

**Comment by bradymiller** [comment]
Looks like the nginx fail is happening about 1 in 3 times, so guessing there is something in 1 (of the 3) facility fixtures that is throwing things off.

**Comment by bradymiller** [comment]
My brain is off or something, but shouldn't this line always fail (since is one rather than greater than 1) in tests/Tests/Services/FacilityServiceTest.php since only of the facilities should be returned (since only one of three has "facility_npi" set to "0123456789" in /tests/Tests/Fixtures/facility.json):
```
        // getAll
        $actualResult = $this->facilityService->getAll(array("facility_npi" => "0123456789"));
        $this->assertNotNull($actualResult);
        $this->assertGreaterThan(1, count($actualResult->getData()));
```

**Comment by yashrajbothra** [comment]
> My brain is off or something, but shouldn't this line always fail (since is one rather than greater than 1) in tests/Tests/Services/FacilityServiceTest.php since only one of the facilities should be returned (since only one of three has "facility_npi" set to "0123456789" in /tests/Tests/Fixtures/facility.json):
> 
> ```
>         // getAll
>         $actualResult = $this->facilityService->getAll(array("facility_npi" => "0123456789"));
>         $this->assertNotNull($actualResult);
>         $this->assertGreaterThan(1, count($actualResult->getData()));
> ```

Ya, Fixtures we have only one fixture of facility_npi with `0123456789`
Let's try with two same npi's :thinking:

**Comment by yashrajbothra** [comment]
Oopsie, Just so there is no id field in the fixture so they are removeFixture will not work which means local will always pass :(

**Comment by bradymiller** [comment]
we got the green light. thank you @yashrajbothra ! bringing this in

**Pull request merged**